[Failing Test]: Various TPC-DS queries throw NPEs using SparkRunner

### What happened?

Various TPC-DS queries started throwing NPEs with the SparkRunner some while back (see [here](http://metrics.beam.apache.org/d/tkqc0AdGk2/tpc-ds-spark-classic-new-sql?orgId=1&from=1689370909359&to=1690703369717)):

```
java.lang.NullPointerException
        at org.apache.beam.vendor.guava.v32_1_2_jre.com.google.common.base.Preconditions.checkNotNull(Preconditions.java:903)
        at org.apache.beam.sdk.util.WindowedValue$TimestampedWindowedValue.<init>(WindowedValue.java:312)
        at org.apache.beam.sdk.util.WindowedValue$TimestampedValueInGlobalWindow.<init>(WindowedValue.java:329)
        at org.apache.beam.sdk.util.WindowedValue.of(WindowedValue.java:95)
```

Without looking further into the underlying root cause, this seems to be related to #27617.

### Issue Failure

Failure: Test is flaky

### Issue Priority

Priority: 2 (backlog / disabled test but we think the product is healthy)

### Issue Components

- [ ] Component: Python SDK
- [ ] Component: Java SDK
- [ ] Component: Go SDK
- [ ] Component: Typescript SDK
- [ ] Component: IO connector
- [ ] Component: Beam examples
- [ ] Component: Beam playground
- [ ] Component: Beam katas
- [ ] Component: Website
- [X] Component: Spark Runner
- [ ] Component: Flink Runner
- [ ] Component: Samza Runner
- [ ] Component: Twister2 Runner
- [ ] Component: Hazelcast Jet Runner
- [ ] Component: Google Cloud Dataflow Runner

getting a different exception on `master` using `--sourceType=CSV` :
`Caused by: java.lang.IllegalArgumentException: Expect 28 fields, but actually 29
        at org.apache.beam.sdk.extensions.sql.impl.schema.BeamTableUtils.csvLines2BeamRows(BeamTableUtils.java:76)
        at org.apache.beam.sdk.tpcds.CsvToRow.lambda$expand$43aa1fdf$1(CsvToRow.java:54)
        at org.apache.beam.sdk.transforms.FlatMapElements$3.processElement(FlatMapElements.java:167)
`

--- a/runners/spark/src/main/java/org/apache/beam/runners/spark/translation/SparkCombineFn.java
+++ b/runners/spark/src/main/java/org/apache/beam/runners/spark/translation/SparkCombineFn.java
@@ -41,7 +41,6 @@ import org.apache.beam.runners.core.construction.SerializablePipelineOptions;
 import org.apache.beam.runners.spark.util.SideInputBroadcast;
 import org.apache.beam.runners.spark.util.SparkSideInputReader;
 import org.apache.beam.sdk.coders.Coder;
-import org.apache.beam.sdk.coders.CoderException;
 import org.apache.beam.sdk.coders.IterableCoder;
 import org.apache.beam.sdk.options.PipelineOptions;
 import org.apache.beam.sdk.transforms.CombineWithContext;
@@ -101,7 +100,7 @@ public class SparkCombineFn<InputT, ValueT, AccumT, OutputT> implements Serializ
         throws Exception;
 
     /**
-     * Merge other acccumulator into this one.
+     * Merge other accumulator into this one.
      *
      * @param other the other accumulator to merge
      */
@@ -173,7 +172,7 @@ public class SparkCombineFn<InputT, ValueT, AccumT, OutputT> implements Serializ
       return new SingleWindowWindowedAccumulator<>(toValue);
     }
 
-    static <InputT, ValueT, AccumT> WindowedAccumulator<InputT, ValueT, AccumT, ?> create(
+    static <InputT, ValueT, AccumT> SingleWindowWindowedAccumulator<InputT, ValueT, AccumT> create(
         Function<InputT, ValueT> toValue, WindowedValue<AccumT> accumulator) {
       return new SingleWindowWindowedAccumulator<>(toValue, accumulator);
     }
@@ -191,10 +190,7 @@ public class SparkCombineFn<InputT, ValueT, AccumT, OutputT> implements Serializ
         Function<InputT, ValueT> toValue, WindowedValue<AccumT> accumulator) {
       this.toValue = toValue;
       this.windowAccumulator = accumulator.getValue();
-      this.accTimestamp =
-          accumulator.getTimestamp().equals(BoundedWindow.TIMESTAMP_MIN_VALUE)
-              ? null
-              : accumulator.getTimestamp();
+      this.accTimestamp = accumulator.getTimestamp();
       this.accWindow = getWindow(accumulator);
     }
 
@@ -247,7 +243,7 @@ public class SparkCombineFn<InputT, ValueT, AccumT, OutputT> implements Serializ
     @Override
     public Collection<WindowedValue<AccumT>> extractOutput() {
       if (windowAccumulator != null) {
-        return Arrays.asList(
+        return Collections.singletonList(
             WindowedValue.of(
                 windowAccumulator, accTimestamp, accWindow, PaneInfo.ON_TIME_AND_ONLY_FIRING));
       }
@@ -516,7 +512,8 @@ public class SparkCombineFn<InputT, ValueT, AccumT, OutputT> implements Serializ
 
     @Override
     public void encode(WindowedAccumulator<InputT, ValueT, AccumT, ?> value, OutputStream outStream)
-        throws CoderException, IOException {
+        throws IOException {
+
       if (type.isMapBased()) {
         wrap.encode(((MapBasedWindowedAccumulator<?, ?, AccumT, ?>) value).map.values(), outStream);
       } else {
@@ -536,7 +533,8 @@ public class SparkCombineFn<InputT, ValueT, AccumT, OutputT> implements Serializ
 
     @Override
     public WindowedAccumulator<InputT, ValueT, AccumT, ?> decode(InputStream inStream)
-        throws CoderException, IOException {
+        throws IOException {
+
       if (type.isMapBased()) {
         return WindowedAccumulator.create(toValue, type, wrap.decode(inStream), windowComparator);
       }

--- a/runners/spark/src/test/java/org/apache/beam/runners/spark/translation/SparkCombineFnTest.java
+++ b/runners/spark/src/test/java/org/apache/beam/runners/spark/translation/SparkCombineFnTest.java
@@ -36,6 +36,7 @@ import org.apache.beam.sdk.transforms.windowing.IntervalWindow;
 import org.apache.beam.sdk.transforms.windowing.PaneInfo;
 import org.apache.beam.sdk.transforms.windowing.Sessions;
 import org.apache.beam.sdk.transforms.windowing.SlidingWindows;
+import org.apache.beam.sdk.transforms.windowing.TimestampCombiner;
 import org.apache.beam.sdk.transforms.windowing.WindowFn;
 import org.apache.beam.sdk.util.CombineFnUtil;
 import org.apache.beam.sdk.util.WindowedValue;
@@ -217,6 +218,34 @@ public class SparkCombineFnTest {
     assertUnorderedEquals(
         Lists.newArrayList("3:999", "5:1999", "5:2999", "2:3999", "1:5999", "1:6999", "1:7999"),
         result);
+  }
+
+  @Test
+  public void testGlobalWindowMergeAccumulatorsWithEarliestCombiner() throws Exception {
+    SparkCombineFn<KV<String, Integer>, Integer, Long, Long> sparkCombineFn =
+        SparkCombineFn.keyed(
+            combineFn,
+            opts,
+            Collections.emptyMap(),
+            WindowingStrategy.globalDefault().withTimestampCombiner(TimestampCombiner.EARLIEST));
+
+    Instant ts = BoundedWindow.TIMESTAMP_MIN_VALUE;
+    WindowedValue<KV<String, Integer>> first = input("key", 1, ts);
+    WindowedValue<KV<String, Integer>> second = input("key", 2, ts);
+    WindowedValue<KV<String, Integer>> third = input("key", 3, ts);
+    WindowedValue<Long> accumulator = WindowedValue.valueInGlobalWindow(0L);
+    SparkCombineFn.SingleWindowWindowedAccumulator<KV<String, Integer>, Integer, Long> acc1 =
+        SparkCombineFn.SingleWindowWindowedAccumulator.create(KV::getValue, accumulator);
+    SparkCombineFn.SingleWindowWindowedAccumulator<KV<String, Integer>, Integer, Long> acc2 =
+        SparkCombineFn.SingleWindowWindowedAccumulator.create(KV::getValue, accumulator);
+    SparkCombineFn.SingleWindowWindowedAccumulator<KV<String, Integer>, Integer, Long> acc3 =
+        SparkCombineFn.SingleWindowWindowedAccumulator.create(KV::getValue, accumulator);
+    acc1.add(first, sparkCombineFn);
+    acc2.add(second, sparkCombineFn);
+    acc3.merge(acc1, sparkCombineFn);
+    acc3.merge(acc2, sparkCombineFn);
+    acc3.add(third, sparkCombineFn);
+    assertEquals(6, (long) Iterables.getOnlyElement(sparkCombineFn.extractOutput(acc3)).getValue());
   }
 
   private static Combine.CombineFn<Integer, Long, Long> getSumFn() {

